src/api.rs

```diff
@@ -0,0 +1,243 @@
+    if member.has_all_permissions() {
+        if let Ok(()) = tokio::fs::remove_file(hub.get_info_path()).await {
+            if let Ok(()) = tokio::fs::remove_dir_all(hub.get_data_path()).await {
+                Ok(())
+            } else {
+                Err(Error::DeleteFailed)
+            }
+        } else {
+            Err(Error::DeleteFailed)
+        }
+    } else {
+        Err(Error::NoPermission)
+    }
+}
+
+pub async fn rename_hub(user: &User, hub_id: &ID, new_name: String) -> Result<String> {
+    is_valid_username(&new_name)?;
+    user.in_hub(&hub_id)?;
+    let mut hub = Hub::load(hub_id).await?;
+    let member = hub.get_member(&user.id)?;
+    if member.has_permission(HubPermission::Administrate, &hub) {
+        let old_name = hub.name.clone();
+        hub.name = new_name;
+        hub.save().await?;
+        Ok(old_name)
+    } else {
+        Err(Error::NoPermission)
+    }
+}
+
+pub async fn change_nickname(user: &User, hub_id: &ID, new_name: String) -> Result<String> {
+    is_valid_username(&new_name)?;
+    user.in_hub(&hub_id)?;
+    let mut hub = Hub::load(hub_id).await?;
+    let member = hub.get_member_mut(&user.id)?;
+    let old_name = member.nickname.clone();
+    member.nickname = new_name;
+    hub.save().await?;
+    Ok(old_name)
+}
+
+pub async fn user_banned(user: &User, hub_id: &ID, user_id: &ID) -> Result<bool> {
+    user.in_hub(hub_id)?;
+    let hub = Hub::load(hub_id).await?;
+    Ok(hub.bans.contains(user_id))
+}
+
+pub async fn user_muted(user: &User, hub_id: &ID, user_id: &ID) -> Result<bool> {
+    user.in_hub(hub_id)?;
+    let hub = Hub::load(hub_id).await?;
+    Ok(hub.mutes.contains(user_id))
+}
+
+pub async fn get_hub_member(user: &User, hub_id: &ID, user_id: &ID) -> Result<HubMember> {
+    user.in_hub(hub_id)?;
+    let hub = Hub::load(hub_id).await?;
+    let member = hub.get_member(&user.id)?;
+    if &user.id == user_id {
+        return Ok(member);
+    } else {
+        hub.get_member(user_id)
+    }
+}
+
+pub async fn join_hub(user: &mut User, hub_id: &ID) -> Result<()> {
+    user.join_hub(hub_id).await?;
+    user.save().await
+}
+
+pub async fn leave_hub(user: &mut User, hub_id: &ID) -> Result<()> {
+    user.leave_hub(hub_id).await?;
+    user.save().await
+}
+
+async fn hub_user_op(user: &User, hub_id: &ID, user_id: &ID, op: HubPermission) -> Result<()> {
+    user.in_hub(hub_id)?;
+    let mut hub = Hub::load(hub_id).await?;
+    let member = hub.get_member(&user.id)?;
+    if member.has_permission(op.clone(), &hub) {
+        match op {
+            HubPermission::Kick => hub.kick_user(user_id).await?,
+            HubPermission::Ban => hub.ban_user(user_id.clone()).await?,
+            HubPermission::Unban => hub.unban_user(user_id),
+            HubPermission::Mute => hub.mute_user(user_id.clone()),
+            HubPermission::Unmute => hub.unmute_user(user_id),
+            _ => return Err(Error::UnexpectedServerArg),
+        }
+        hub.save().await
```

Comment:
It might be better to have the data in memory or controlled by a database rather than storing these in JSON files which you need to read and write for each request.

Reply:
By that do you mean use a database for storing data such as hub info (name, id, users, groups, etc..)?

Reply:
I thought about using dedicated database software but had a hard time finding one that did not require external software or that had an easily typed api...